Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+    return true;
+  };
+
+  auto& currentBuffer = getCurrentBuffer();
+  auto currentBufId = getCurrentBufId();
+  auto& previousBuffer = getPreviousBuffer();
+  auto previousBufId = getPreviousBufId();
+
+  if (mPrint) {
+    std::cout << "Merging digits in " << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < previousBuffer.digits.size(); i++) {
+    MergerDigit& d1 = previousBuffer.digits[i];
+
+    // skip digits that do not start at the beginning of the time window
+    Digit::Time startTime = d1.digit.getTime();
+    if (startTime.sampaTime != 0) {
+      continue;
+    }
+
+    for (size_t j = 0; j < previousBuffer.digits.size(); j++) {
+      if (i == j) {
+        continue;
+      }
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+
+  // only merge digits from consecutive orbits
+  uint32_t orbit_p = previousBuffer.orbit;
+  uint32_t orbit_c = currentBuffer.orbit;
+  if (mPrint) {
+    std::cout << "orbit_c: " << orbit_c << "  orbit_p: " << orbit_p << std::endl;
+  }
+  if ((orbit_c >= orbit_p) && ((orbit_c - orbit_p) > 1)) {
+    return;
+  }
+
+  if (mPrint) {
+    std::cout << "Merging digits from " << currentBufId << " (orbit=" << currentBuffer.orbit << ") into "
+              << previousBufId << " (orbit=" << previousBuffer.orbit << ")\n";
+  }
+
+  for (size_t i = 0; i < currentBuffer.digits.size(); i++) {
+    MergerDigit& d1 = currentBuffer.digits[i];
+
+    // skip digits that do not start at the beginning of the time window
+    Digit::Time startTime = d1.digit.getTime();
+    if (startTime.sampaTime != 0) {
+      continue;
+    }
+
+    for (size_t j = 0; j < previousBuffer.digits.size(); j++) {
+      MergerDigit& d2 = previousBuffer.digits[j];
+      if (checkDigits(d1, d2)) {
+        break;
+      }
+    }
+  }
+}
+
+void Merger::setOrbit(int feeId, uint32_t orbit, bool stop)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].setOrbit(orbit, stop);
+}
+
+void Merger::setDigitHandler(std::function<void(const Digit&)> h)
+{
+  for (int feeId = 0; feeId <= MCH_MERGER_FEEID_MAX; feeId++) {
+    mergers[feeId].setDigitHandler(h);
+  }
+}
+
+void Merger::addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                      int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].getCurrentBuffer().digits.emplace_back(MergerDigit{o2::mch::Digit(deId, padId, adc, time),
+                                                                    stopTime, false, solarId, dsAddr, chAddr});
+}
+
+void Merger::mergeDigits(int feeId)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].mergeDigits();
+}
+
+//=======================
+// Data decoder
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
```

Comment:
this function is quite long and difficult to follow as it is now

Reply:
You are right, I will try to improve it. However I am not sure we can make the channel handler much shorter...

Reply:
Agreed. For now, and for the channelHandler at least ;-) (as I don't have immediate proposal on how to improve it) 

I still believe though that 4 lambdas in this method is a bit much and makes the code quite hard to read.  Probably 2/4 could be externalised : e.g. the dumpDigits could be a simple free function take a vector of digits as argument, and the updateMerger, well, I'd have to think about it ;-) )